my code...

Volder5 opened the pull request:
got some code here to show

Commit:
my code...

Commit:
More security

Commit:
Security update...

roman-sachenko commented on .DS_Store:
This file should be ignored. Read (and learn) about ".gitignore".

Volder5 replied:
done

roman-sachenko commented on config/settings.py:
```diff
@@ -66,16 +70,20 @@
     'default': {
-        'ENGINE': 'django.db.backends.sqlite3',
-        'NAME': BASE_DIR / 'db.sqlite3',
+        'ENGINE': 'django.db.backends.postgresql',
```
You should use environment variables for storing configurations. What you're doing here is not secure. Read about best practices for setting configurations and environment variables for Python.

Volder5 replied:
Done

roman-sachenko commented on config/settings.py:
```diff
@@ -120,3 +128,14 @@
 
 DEFAULT_AUTO_FIELD = 'django.db.models.BigAutoField'
+
+EMAIL_BACKEND = 'django.core.mail.backends.smtp.EmailBackend'
```
Same as before, all the configurations should come from environment (environment variables)

Volder5 replied:
Done

roman-sachenko reviewed (commented):
Good start. 👍  A few comments below.

- [ ] Missing RERADME.md file (read about README.md); It should contain the project description, instructions on how to set up, run it locally, and in production.
- [ ] You should dockerize your applications (read and learn about Docker, Dockerfile)
- [ ] Please add a startup script(s) that can be run to start the application locally and in production. There should be separate commands for each.
- [ ] I'd recommend correcting the way you name GH (GitHub) branches and commits; Read about Git Convention: https://www.conventionalcommits.org
- [ ] This pull request doesn't contain any logic or description. Please provide exactly what is coming in this Pull Request in its description 

Read:
- https://www.geeksforgeeks.org/how-to-manage-local-vs-production-settings-in-django/
- https://www.conventionalcommits.org
- https://github.com/jehna/readme-best-practices

roman-sachenko commented on apps/users/views.py:
```diff
@@ -0,0 +1,75 @@
+            messages.error(request, "Invalid credentials")
+    return render(request, "users/login.html")
+
+import random
```
Place all the imports at the top of the page, so you don't scatter them within your scripts.

Volder5 replied:
Done

roman-sachenko commented on apps/users/views.py:
```diff
@@ -0,0 +1,75 @@
+
+
+def login_page(request):
+    if request.method == "POST":
```
This is not clear. It should specifically state which HTTP method is used for the route. Find this for your framework and explicitly state that you are listening to POST only. You should NOT have to manually check this.

Volder5 replied:
Done

roman-sachenko commented on apps/users/views.py:
```diff
@@ -0,0 +1,75 @@
+def login_page(request):
+    if request.method == "POST":
+        username = request.POST.get('username')
+        password = request.POST.get('password')
```
Validation is missing. How do you know that the user didn't send any corrupted or malicious input? You should validate your parameters. Read about input data validation in your framework.

Volder5 replied:
Done

roman-sachenko commented on apps/users/views.py:
```diff
@@ -0,0 +1,75 @@
+from django.core.mail import send_mail
+
+def register_page(request):
+    if request.method == "POST":
```
The same as above. You should not have a manual check for this. This should be done in routes; an explicit HTTP method should be provided for the route.

Volder5 replied:
Done

roman-sachenko commented on apps/users/views.py:
```diff
@@ -0,0 +1,75 @@
+
+def register_page(request):
+    if request.method == "POST":
+        username = request.POST.get('username')
```
Parameters validation is missing. Read up on how to perform input parameters validation.

Volder5 replied:
Done

roman-sachenko commented on apps/users/views.py:
```diff
@@ -0,0 +1,75 @@
+    return render(request, "users/register.html")
+
+def email_verification_page(request):
+    if request.method == "POST":
```
Fix this in routes. Same as before.

Volder5 replied:
dOne

roman-sachenko commented on apps/users/views.py:
```diff
@@ -0,0 +1,75 @@
+
+        if input_code == reg_data.get('code'):
+            User.objects.create_user(
+                username=reg_data['username'],
```
Input data validation is missing

Volder5 replied:
Done

roman-sachenko commented on apps/users/views.py:
```diff
@@ -0,0 +1,75 @@
+            User.objects.create_user(
+                username=reg_data['username'],
+                email=reg_data['email'],
+                password=reg_data['password']
```
Continue reading on how to store passwords. You are not supposed to store passwords in raw format. They should be hashed.

Volder5 replied:
They are already hashed by django

Volder5 replied:
automatic

roman-sachenko commented on apps/users/views.py:
```diff
@@ -0,0 +1,75 @@
+                'code': verification_code
+            }
+
+            send_mail(
```
How does send_mail function work? Have you checked it? Does it work now?

Volder5 replied:
Yeah, completely work

roman-sachenko commented on apps/users/views.py:
```diff
@@ -0,0 +1,75 @@
+            request.session['registration_data'] = {
+                'username': username,
+                'email': email,
+                'password': password,
```
It's not safe to store passwords 'as is' in sessions. Also, why do you need the password stored in the session?

Volder5 replied:
now i made that it works in way that it's creating "EmailVerification" module where creates temporary data with valid key. no sessions now

Volder5 replied:
model*

roman-sachenko commented on apps/users/models.py:
```diff
@@ -0,0 +1,4 @@
+from django.db import models
```
Where is the user model? I see that you have functionality to create users, but you don't have its model. How does the DB (Database Connection) work?

Volder5 replied:
Hell na I don't know by myself, thats frameworks model. Im gonna check more about it and after maybe change it to my own

Volder5 replied:
I checked little bit about it, User is involved in django from start and other tables you can create in "models.py" and work with them

Commit:
Added .env for securiiity

Commit:
Added .env for securiiity 2

Commit:
fix(auth): Security updates and some features and 'answers' to @roman-sachenko's comments.
Now it's secure from sql injections and have some check on data that user sends to the server

Commit:
fix/feature(auth): Made if user tried to registrate but left email verification page and try to do registrate with same data now it works

Commit:
feat(auth): Recovery function

Commit:
fix(auth): Fixed error messages from start

Commit:
fix(auth): now request method check is more clean and not combined with buissnes logic as it was before

Commit:
fix(auth): Fixed bug in app/users/urls.py, now 'recovery' is not required and started with docker(not done yet)

Commit:
feat(auth): Yoy, I started to making profile with some functional, also it has some bugfixes in this commit

Commit:
feature(auth): Now when user is verified that creates profile for him (not ideal logic but I will check more about it)

Commit:
feature(auth): Now profile is more dynamic

roman-sachenko commented on Dockerfile:
Fix service setup. Whenever you start a service, it should perform a self-check, so it should check that all the configurations are made. If this is not the case, you need to show a proper error message with a hint on how to fix the problem. Something like "Environment configurations are not set, please do this and that."

I get this error when I try to run a service in docker.


```
docker run -p 8000:8000 60fda692238ac3303bc7e93aa501c4237b1e0e78768c44d4881bef48072b4069
Traceback (most recent call last):
  File "/app/manage.py", line 22, in <module>
    main()
  File "/app/manage.py", line 18, in main
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.11/site-packages/django/core/management/__init__.py", line 442, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.11/site-packages/django/core/management/__init__.py", line 382, in execute
    settings.INSTALLED_APPS
  File "/usr/local/lib/python3.11/site-packages/django/conf/__init__.py", line 81, in __getattr__
    self._setup(name)
  File "/usr/local/lib/python3.11/site-packages/django/conf/__init__.py", line 68, in _setup
    self._wrapped = Settings(settings_module)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/django/conf/__init__.py", line 166, in __init__
    mod = importlib.import_module(self.SETTINGS_MODULE)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/app/config/settings.py", line 23, in <module>
    SECRET_KEY = config("SECRET_KEY")
                 ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/decouple.py", line 248, in __call__
    return self.config(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/decouple.py", line 107, in __call__
    return self.get(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/decouple.py", line 92, in get
    raise UndefinedValueError('{} not found. Declare it as envvar or define a default value.'.format(option))
decouple.UndefinedValueError: SECRET_KEY not found. Declare it as envvar or define a default value.

```

roman-sachenko commented on README.md:
```diff
@@ -0,0 +1 @@
+thats a "readme" file, but I didn't made enough for making "readme"
```
Fix README. It doesn't contain information about your service, instructions on how to run it locally, etc. But it should.

roman-sachenko commented on apps/main/migrations/0001_initial.py:
```diff
@@ -0,0 +1,27 @@
+                ('id', models.BigAutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
+                ('bio', models.TextField(blank=True, max_length=50)),
+                ('profile_picture', models.ImageField(blank=True, upload_to='profile_pics/')),
+                ('location', models.CharField(blank=True, max_length=30)),
```
What location is referred to? Is this a country or full address? If we talk about a country, then I'd recommend using country codes. Read up on country codes. For them, you need to use enumerations to prevent random values from being set. Read about country codes, enumerations in Python.

roman-sachenko commented on config/settings.py:
```diff
@@ -20,10 +20,10 @@
 
 # SECURITY WARNING: keep the secret key used in production secret!
-SECRET_KEY = 'django-insecure-2hnjr4%jaye3n3kfx5l@n7fhrrfe(=t(z9y^8gl-c+07&4kpus'
+SECRET_KEY = config("SECRET_KEY")
```
What is this secret key? Which service does it represent? If it corresponds to a specific service, it would be beneficial to specify it in the variable name. For example, GOOGLE_ACCESS_TOKEN, STRIPE_ACCESS_TOKEN, and so on.

roman-sachenko commented on apps/users/forms.py:
```diff
@@ -0,0 +1,86 @@
+        email = self.cleaned_data['email']
+        if User.objects.filter(email=email, is_active=True).exists():
+            raise ValidationError("This email is already registered.")
+        if len(email) > 50:
```
I don't agree that this is good validation. What you should do is check if the input text is actually an email. Look for libraries in Python for this.

roman-sachenko commented on apps/users/forms.py:
```diff
@@ -0,0 +1,86 @@
+        return email
+    
+class PasswordForm(forms.Form):
+    password = forms.CharField(max_length=config("PASSWORD_MAX_LENGTHT"), required=True, widget=forms.PasswordInput)
```
Why do you need PASSWORD_MAX_LENGTHT?

roman-sachenko commented on apps/users/forms.py:
```diff
@@ -0,0 +1,86 @@
+        password = cleaned_data.get("password")
+        password2 = cleaned_data.get("password2")
+
+        if password != password2:
```
Why do you need this on the back-end? There's no point in comparing passwords once they're passed to the front-end, especially if we're talking about the user creation process. It's enough to simply accept the password, that's it.

roman-sachenko commented on apps/users/migrations/0002_profile.py:
```diff
@@ -0,0 +1,26 @@
+                ('id', models.BigAutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
+                ('bio', models.TextField(blank=True, max_length=50)),
+                ('profile_picture', models.ImageField(blank=True, null=True, upload_to='profile_pics/')),
+                ('joined_at', models.DateField(auto_now_add=True)),
```
Do you have roles? Like regular user, admin?

roman-sachenko reviewed (changes requested):
I couldn't start the application. Please update the README and provide detailed instructions on how I can start it up in Docker, so I have my database in Docker, service in Docker, and then I can play around and use the application.

To ensure that the instructions work, I recommend cloning the project in a different directory on your local machine and running it from scratch by following the instructions. Then please notify me and I'll try things out from my end.

TIPS:
- Keep Pull Requests small, yours is too big. Usually, it's not okay to have such large pull requests. That's why I recommended to you before to roll out features gradually.
- Always have README up to date, which contains information about application setup, description, how to start locally, in production, etc.
- Find a proper data/input validation library for Python, so you can validate emails, names etc. 
- Format PR (Pull Request) description. "Got some code here to show" is not a suitable description, unfortunately, since it doesn't provide any insights about the features and functionality I am about to review here. Lookup for examples on the Internet (https://github.com/codica2/pull-request-best-practices/blob/master/.gitlab/merge_request_templates/Task.md)
- Work on your commit messages. They should not be long but should also provide a hint about what is occurring. They should be in the present tense, not the past (use "make" instead of "made").
- Follow branch name convention: https://medium.com/@abhay.pixolo/naming-conventions-for-git-branches-a-cheatsheet-8549feca2534

Good README example: 
https://github.com/matiassingers/awesome-readme

Useful links: 
- https://github.com/matiassingers/awesome-readme
- https://medium.com/@abhay.pixolo/naming-conventions-for-git-branches-a-cheatsheet-8549feca2534
